File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        n = len(nums)
        mp = {}
        for e in nums:
            mp[e] = mp.get(e , 0)+1
        ans = [[] for _ in range(n + 1)] 
        for key , val in mp.items():
            ans[val].append(key)
        
        res = []
        for i in range(n , -1  , -1):
            for j in ans[i]:
                if k > 0:
                    res.append(j)
                    k -= 1
                    if k == 0:break
        return res

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        HashMap<Integer , Integer> mp = new HashMap<>();
        int n = nums.length;
        ArrayList<Integer>[] ans = new ArrayList[n + 1];
        for(int i = 0 ; i < n ; i++){
            mp.put(nums[i] , mp.getOrDefault(nums[i] , 0) + 1);
        }
        for(Map.Entry<Integer , Integer>entry : mp.entrySet()){
            int val = entry.getKey();
            int idx = entry.getValue();
            if(ans[idx] == null){
                ans[idx] = new ArrayList<>();
            }
            ans[idx].add(val);
        }
        int[] res = new int[k];
        int idx = 0;
        for(int i = n ; i >= 0 && idx < k ;i--){
            if(ans[i] != null){
                for(int p : ans[i]){
                    res[idx++] = p;
                    if(idx == k) break;
                }
            }
        }
        return res;
    }
}